routes/post.go

```diff
@@ -1599,17 +1600,601 @@ func (fes *APIServer) GetDiamondedPosts(ww http.ResponseWriter, req *http.Reques
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("GetDiamondsForPost: Error constucting utxoView: %v", err))
+		return
+	}
+
+	// Fetch the diamonds for the post requested.
+	pkidToDiamondLevel, err := utxoView.GetDiamondSendersForPostHash(postHash)
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("GetDiamondsForPost: Error getting pkidToDiamondLevel map %v", err))
+		return
+	}
+
+	// Filter out any restricted profiles.
+	pkMapToFilter := make(map[lib.PkMapKey][]byte)
+	for senderPKID := range pkidToDiamondLevel {
+		profileEntry := utxoView.GetProfileEntryForPKID(&senderPKID)
+		if profileEntry != nil {
+			pkMapKey := lib.MakePkMapKey(profileEntry.PublicKey)
+			pkMapToFilter[pkMapKey] = profileEntry.PublicKey
+		}
+	}
+	filteredPkMap, err := fes.FilterOutRestrictedPubKeysFromMap(pkMapToFilter, readerPublicKeyBytes, "leaderboard" /*moderationType*/)
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("GetDiamondsForPost: Error filtering out restricted profiles: %v", err))
+		return
+	}
+
+	// Create a list of unifltered (aka not blacklisted) diamondSenders.
+	diamondSenders := []*lib.ProfileEntry{}
+	for senderPKID := range pkidToDiamondLevel {
+		profileEntry := utxoView.GetProfileEntryForPKID(&senderPKID)
+		if profileEntry == nil {
+			continue
+		} else {
+			if _, ok := filteredPkMap[lib.MakePkMapKey(profileEntry.PublicKey)]; ok {
+				diamondSenders = append(diamondSenders, profileEntry)
+			}
+		}
+	}
+
+	// Almost done. Just need to sort the comments.
+	sort.Slice(diamondSenders, func(ii, jj int) bool {
+
+		// Attempt to sort on bitclout locked.
+		iiBitCloutLocked := diamondSenders[ii].BitCloutLockedNanos
+		jjBitCloutLocked := diamondSenders[jj].BitCloutLockedNanos
+		if iiBitCloutLocked > jjBitCloutLocked {
+			return true
+		} else if iiBitCloutLocked < jjBitCloutLocked {
+			return false
+		}
+
+		// Attempt to sort on diamond level.
+		iiPKID := utxoView.GetPKIDForPublicKey(diamondSenders[ii].PublicKey)
+		jjPKID := utxoView.GetPKIDForPublicKey(diamondSenders[jj].PublicKey)
+		iiDiamondLevel := pkidToDiamondLevel[*iiPKID.PKID]
+		jjDiamondLevel := pkidToDiamondLevel[*jjPKID.PKID]
+		if iiDiamondLevel > jjDiamondLevel {
+			return true
+		} else if iiDiamondLevel < jjDiamondLevel {
+			return false
+		}
+
+		// Sort based on pub key if all else fails.
+		return lib.PkToString(diamondSenders[ii].PublicKey, fes.Params) > lib.PkToString(diamondSenders[jj].PublicKey, fes.Params)
+	})
+
+	// Cut out the page of diamondSenders that we care about.
+	diamondSendersLength := uint32(len(diamondSenders))
+	// Slice the comments from the offset up to either the end of the slice or the offset + limit, whichever is smaller.
+	maxIdx := lib.MinUint32(diamondSendersLength, requestData.Offset+requestData.Limit)
+	var diamondSendersPage []*lib.ProfileEntry
+	if diamondSendersLength > requestData.Offset {
+		diamondSendersPage = diamondSenders[requestData.Offset:maxIdx]
+	}
```

Comment:
this appears incorrect to me.  I think we always want to take a subset of the slice from `[requestData.Offset:maxIdx]`, not only when Offset is greater than diamondSendersLength?